pkg/beaconfqdn/mongodb.go

```diff
@@ -151,8 +178,189 @@ func (r *repo) Upsert(hostnameMap map[string]*hostname.Input, minTimestamp, maxT
+	var affectedHostnameIPMap map[string][]data.UniqueIP = make(map[string][]data.UniqueIP, len(hostMap)/10)
+
+	for _, host := range hostMap {
+		if host.IsLocal {
+			continue
+		}
+
+		externalHosts = append(externalHosts, host.Host)
+
+		if len(externalHosts) == 200000 { // we've hit the limit, run the MongoDB query
+			reverseDNSagg := reverseDNSQueryWithIPs(externalHosts)
+			externalHosts = externalHosts[:0] // clear externalHosts to make room for the next chunk
+
+			affectedHostnamesBuffer = nil
+			err := ssn.DB(r.database.GetSelectedDB()).C(r.config.T.DNS.HostnamesTable).
+				Pipe(reverseDNSagg).AllowDiskUse().All(&affectedHostnamesBuffer)
+			if err != nil {
+				return []hostnameIPs{}, err
+			}
+
+			for i := range affectedHostnamesBuffer {
+				affectedHostnameIPMap[affectedHostnamesBuffer[i].Host] = affectedHostnamesBuffer[i].ResolvedIPs
+			}
+		}
+	}
+
+	if len(externalHosts) > 0 { // take care of running the MongoDB query for any leftover hosts
+		reverseDNSagg := reverseDNSQueryWithIPs(externalHosts)
+		externalHosts = nil // nolint , we are assigning nil here to allow the GC to free this potentially large array
+
+		affectedHostnamesBuffer = nil
+		err := ssn.DB(r.database.GetSelectedDB()).C(r.config.T.DNS.HostnamesTable).
+			Pipe(reverseDNSagg).AllowDiskUse().All(&affectedHostnamesBuffer)
+		if err != nil {
+			return []hostnameIPs{}, err
+		}
+
+		for i := range affectedHostnamesBuffer {
+			affectedHostnameIPMap[affectedHostnamesBuffer[i].Host] = affectedHostnamesBuffer[i].ResolvedIPs
+		}
+	}
+
+	// convert the map into a slice
+	affectedHostnamesBuffer = make([]hostnameIPs, 0, len(affectedHostnameIPMap))
+	for host, ips := range affectedHostnameIPMap {
+		affectedHostnamesBuffer = append(affectedHostnamesBuffer, hostnameIPs{
+			Host:        host,
+			ResolvedIPs: ips,
+		})
+	}
+	return affectedHostnamesBuffer, nil
+}
+
+/*
+db.getCollection('hostnames').aggregate([
+	{"$match": { "$or": [
+		{
+			"dat.ips.ip": "104.16.107.25",
+			"dat.ips.network_uuid": UUID("ffffffff-ffff-ffff-ffff-ffffffffffff"),
+		},
+		{
+			"dat.ips.ip" : "104.16.229.152",
+			"dat.ips.network_uuid" : UUID("ffffffff-ffff-ffff-ffff-ffffffffffff"),
+		},
+	]}},
+	{"$project": {
+		"host": 1,
+		"dat.ips.ip": 1,
+		"dat.ips.network_uuid": 1,
+	}},
+	{"$unwind": "$dat"},
+	{"$unwind": "$dat.ips"},
+	{"$group": {
+		"_id": {
+			"host": "$host",
+			"ip": "$dat.ips.ip",
+			"network_uuid": "$dat.ips.network_uuid",
+		},
+	}},
+	{"$group": {
+		"_id": "$_id.host",
+		"ips": {"$push": {
+			"ip": "$_id.ip",
+			"network_uuid": "$_id.network_uuid",
+		}},
+	}}
+])
+
+reverseDNSQueryWithIPs returns a MongoDB aggregation which returns the hostnames associated with the given
+UniqueIPs. Additionally, all of the IPs associated with each hostname are returned.
+*/
+func reverseDNSQueryWithIPs(uniqueIPs []data.UniqueIP) []bson.M {
+	var uniqueIPBsonSelectors []bson.M = make([]bson.M, 0, len(uniqueIPs))
```

Comment:
Not a big deal and will just comment in this one place, but I don't think the "var" and data type aren't needed on the left with the "make" on the right.